File: 1.3/BinaryNotesMQ/java/src/org/bn/mq/net/tcp/Transport.java
/*
 * Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
 * 
 * Licensed under the LGPL, Version 2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.gnu.org/copyleft/lgpl.html
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 * With any your questions welcome to my e-mail 
 * or blog at http://abdulla-a.blogspot.com.
 */

package org.bn.mq.net.tcp;

import java.io.IOException;

import java.net.URI;

import java.nio.ByteBuffer;
import java.nio.channels.SocketChannel;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import org.bn.mq.IMessage;

import org.bn.mq.net.ITransport;
import org.bn.mq.net.ITransportListener;
import org.bn.mq.net.ITransportMessageCoder;
import org.bn.mq.protocol.MessageEnvelope;

public abstract class Transport implements ITransport {
    private URI addr;
    private SocketChannel socket = null;
    protected ReadWriteLock socketLock =  new ReentrantReadWriteLock();
    protected List<ITransportListener> listeners = new LinkedList<ITransportListener>();
    protected SocketFactory socketFactory;
    protected ByteBuffer tempReceiveBuffer = ByteBuffer.allocate(64*1024);
    protected ITransportMessageCoder messageCoder;
    protected final Lock callLock = new ReentrantLock();
    protected final Condition callLockEvent  = callLock.newCondition();    
    protected String currentCallMessageId = "-none-";
    protected MessageEnvelope currentCallMessage = null;
    
    public Transport(URI addr, SocketFactory factory) {
        setAddr(addr);
        this.socketFactory = factory ;
    }
       
    protected void onTransportClosed() {
        close();
    }    
    
    public void close() {
        socketLock.writeLock().lock();
        if(isAvailable()) {
            try {
                getSocket().close();
                setSocket(null);
            }
            catch (IOException e) {
                // TODO
            }
        }
        socketLock.writeLock().unlock();
    }

    
    public ByteBuffer receiveAsync() throws IOException {
        socketLock.readLock().lock();
        SocketChannel channel = getSocket();
        ByteBuffer result = null;
        try {
            if(isAvailable()) {
                result = ByteBuffer.allocate(4096);                
                int readedBytes = -1;
                do {                                        
                    tempReceiveBuffer.clear();
                    readedBytes = channel.read(tempReceiveBuffer);
                    if(readedBytes>0) {
                        if(result.remaining()<readedBytes) {
                            byte[] data = result.array();
                            result = ByteBuffer.allocate(data.length+4096);
                            result.put(data);
                        }
                        result.put(tempReceiveBuffer.array(),result.position(),tempReceiveBuffer.position());
                    }
                    else
                    if(readedBytes==-1) {
                        onTransportClosed();
                    }                    
                }                
                while(readedBytes>0);
                result.flip();
            }
            else {
                throw new IOException("Not connected");
            }
        }
        finally {
            socketLock.readLock().unlock();
        }    
        return result;        
    }
    
    public void sendAsync(ByteBuffer buffer) throws IOException {
        if(socketFactory.getWriterStorage()!=null) {
            if(isAvailable()) {
                socketFactory.getWriterStorage().pushPacket(this, buffer);
            }
            else
                throw new IOException("Transport is not connected!");
        }
        else
            throw new IOException("Unable to write readonly transport!");
    }
    
    public void sendAsync(byte[] buffer) throws IOException {
        sendAsync(ByteBuffer.wrap(buffer));
    }
    
    public void send(MessageEnvelope message) throws Exception {
        ByteBuffer buffer = messageCoder.encode(message);
        send(buffer);
    }
        
    public void sendAsync(MessageEnvelope message) throws Exception {
        ByteBuffer buffer = messageCoder.encode(message);
        sendAsync(buffer);
    }     
    
    public void send(ByteBuffer buffer) throws IOException {
        socketLock.readLock().lock();        
        try {                        
            if(isAvailable()) {
                SocketChannel channel = getSocket();
                if(channel!=null)
                    channel.write(buffer);
            }
            else {
                throw new IOException("Not connected");
            }
        }
        finally {
            socketLock.readLock().unlock();
        }    
    }

    public void send(byte[] buffer) throws IOException {
        send(ByteBuffer.wrap(buffer));
    }
    
    public void addListener(ITransportListener listener) {
        synchronized(listeners) {
            listeners.add(listener);
        }
    }

    public void delListener(ITransportListener listener) {
        synchronized(listeners) {
            listeners.remove(listener);
        }    
    }
    
    protected void fireConnectedEvent() {
        synchronized(listeners) {
            for(ITransportListener listener: listeners) {                    
                listener.onConnected(this);
            }            
        }
    }

    protected void fireDisconnectedEvent() {
        synchronized(listeners) {
            for(ITransportListener listener: listeners) {                    
                listener.onDisconnected(this);
            }            
        }
    }
    
    protected void fireReceivedData(ByteBuffer packet) throws Exception {
        synchronized(listeners) {
            MessageEnvelope message = messageCoder.decode(packet);
            boolean doProcessListeners = true;            
            if(message!=null) {
                callLock.lock();
                if(currentCallMessageId.equals(message.getId())) {
                    currentCallMessage = message;
                    callLockEvent.signal();
                    doProcessListeners = false;
                }
                callLock.unlock();
            }
            if(doProcessListeners) {
                for(ITransportListener listener: listeners) {                    
                    listener.onReceive(message,this);
                }
            }
        }
    }

    public synchronized MessageEnvelope call(MessageEnvelope message, int timeout) throws Exception {
        MessageEnvelope result = null;
        callLock.lock();
        currentCallMessage = null;
        currentCallMessageId = message.getId();        
        try {
            sendAsync(message);
            callLockEvent.await(timeout,TimeUnit.SECONDS);
            result = currentCallMessage;
            currentCallMessageId = "";
        }
        finally {
            callLock.unlock();
        }
        return result;    
    }
    
    public synchronized MessageEnvelope call(MessageEnvelope message) throws Exception {
        return this.call(message,120); // By default timeout is 2 min
    }

    public URI getAddr() {
        return addr;
    }

    protected void setAddr(URI addr) {
        this.addr = addr;
    }

    public SocketChannel getSocket() {
        return socket;
    }

    public void setSocket(SocketChannel socket) {
        socketLock.writeLock().lock();
        this.socket = socket;
        if(this.socketFactory!=null) {
            if(this.socket!=null) {
                this.socketFactory.getReaderStorage().addTransport(this);
                messageCoder = socketFactory.getTransportFactory().getTransportMessageCoderFactory().newCoder(this);
            }
            else {
                this.socketFactory.getReaderStorage().removeTransport(this);
            }
        }
        socketLock.writeLock().unlock();
    }

    public boolean isAvailable() {
        socketLock.readLock().lock();
        boolean result = getSocket() != null && getSocket().isOpen() && getSocket().isConnected() ;
        socketLock.readLock().unlock();
        return result;
    }
    
    public void finalize() {
        close();
    }
    
    
}

File: 1.3/BinaryNotesMQ/.net/BinaryNotesMQ/src/org/bn/mq/net/tcp/Transport.cs
/*
* Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
* 
* Licensed under the LGPL, Version 2 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* 
*      http://www.gnu.org/copyleft/lgpl.html
* 
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* 
* With any your questions welcome to my e-mail 
* or blog at http://abdulla-a.blogspot.com.
*/
using System;
using System.Threading;
using System.Collections.Generic;
using org.bn.mq.protocol;

namespace org.bn.mq.net.tcp
{
	
	public abstract class Transport : ITransport
	{
		protected Uri addr;
		private String socket = null;
		//protected internal ReadWriteLock socketLock = new ReentrantReadWriteLock();
		protected internal SocketFactory socketFactory;
		
		//protected internal ReadWriteLock listenersLock = new ReentrantReadWriteLock();
		protected IList < ITransportConnectionListener > listeners = new List < ITransportConnectionListener >();
		
        //protected internal ReadWriteLock readersLock = new ReentrantReadWriteLock();
		protected IList < ITransportReader > readers = new List < ITransportReader >();
		
		protected internal ITransportReader unhanledReader = null;
		
		protected internal ByteBuffer tempReceiveBuffer = ByteBuffer.allocate(64 * 1024);
		protected internal ITransportMessageCoder messageCoder;
		
		protected internal AutoResetEvent callLockEvent = new AutoResetEvent(false);
		protected string currentCallMessageId = "-none-";
		protected internal MessageEnvelope currentCallMessage = null;
		
		public Transport(Uri addr, SocketFactory factory)
		{
			setAddr(addr);
			this.socketFactory = factory;
		}
		
		protected internal virtual void  onTransportClosed()
		{
			close();
		}

		virtual public void setUnhandledMessagesReader(ITransportReader value)
		{
			//readersLock.writeLock().lock_Renamed();
            lock(readers)
            {
				this.unhanledReader = value;
            }
			//readersLock.writeLock().unlock();			
		}

        public String getSocket() 
        {
            return null;
        }

        public void setSocket(String socket)
        {
			/*set
			{
				socketLock.writeLock().lock_Renamed();
				this.socket = value;
				if (this.socketFactory != null)
				{
					if (this.socket != null)
					{
						this.socketFactory.ReaderStorage.addTransport(this);
						this.socketFactory.WriterStorage.addAliveReqInspection(this);
						messageCoder = socketFactory.getTransportFactory().TransportMessageCoderFactory.newCoder(this);
					}
					else
					{
						this.socketFactory.ReaderStorage.removeTransport(this);
						this.socketFactory.WriterStorage.delAliveReqInspection(this);
					}
				}
				socketLock.writeLock().unlock();
			}*/
        }


		
		public virtual void  close()
		{
            lock(addr) // Compact Framework doesn't supported RW-Mutex & Semaphores :(
            {
			    if (isAvailable())
			    {
				    try
				    {
					    //Socket.close();
					    setSocket(null);
				    }
				    catch (System.IO.IOException e)
				    {
					    // TODO
				    }
			    }
            }
		}
		
		
		public virtual ByteBuffer receiveAsync()
		{
			ByteBuffer result = null;
			if (isAvailable())
			{
				result = ByteBuffer.allocate(4096);
				int readedBytes = - 1;
				do 
				{
                    lock(addr) // Compact Framework doesn't supported RW-Mutex & Semaphores :(
                    {
					    tempReceiveBuffer.clear();
					    /*SocketChannel channel = Socket;
					    if (channel != null)
						    readedBytes = channel.read(tempReceiveBuffer);
                         */
					    if (readedBytes > 0)
					    {
						    if (result.Remaining < readedBytes)
						    {
							    byte[] data = result.Value;
							    result = ByteBuffer.allocate(data.Length + 4096);
							    result.put(data);
						    }
						    result.put(tempReceiveBuffer.Value, result.Position, tempReceiveBuffer.Position);
					    }
					    if (readedBytes == - 1)
					    {
						    onTransportClosed();
					    }
                    }   
			    }
			    while (readedBytes > 0);
			    result.flip();
			}
			else
			{
				throw new System.IO.IOException("Not connected");
			}
			return result;
		}
		
		public virtual void  sendAsync(ByteBuffer buffer)
		{
			if (socketFactory.WriterStorage != null)
			{
				if (isAvailable())
				{
					socketFactory.WriterStorage.pushPacket(this, buffer);
				}
				else
					throw new System.IO.IOException("Transport is not connected!");
			}
			else
				throw new System.IO.IOException("Unable to write readonly transport!");
		}
		
		public virtual void  sendAsync(byte[] buffer)
		{
			sendAsync(ByteBuffer.wrap(buffer));
		}
		
		public virtual void  send(MessageEnvelope message)
		{
			if (isAvailable())
			{
				ByteBuffer buffer = messageCoder.encode(message);
				send(buffer);
			}
			else
				throw new System.IO.IOException("Transport is not connected!");
		}
		
		public virtual void  sendAsync(MessageEnvelope message)
		{
			if (isAvailable())
			{
				ByteBuffer buffer = messageCoder.encode(message);
				sendAsync(buffer);
			}
			else
				throw new System.IO.IOException("Transport is not connected!");
		}
		
		public virtual void  send(ByteBuffer buffer)
		{
			try
			{
				lock(addr) 
                {
				    if (isAvailable())
				    {
					    /*SocketChannel channel = Socket;
					    if (channel != null)
						    channel.write(buffer);
                         */
				    }
				    else
				    {
					    throw new System.IO.IOException("Not connected");
				    }
                }
			}
			catch (System.Exception ex)
			{
				this.onTransportClosed();
			}
		}
		
		public virtual void  send(byte[] buffer)
		{
			send(ByteBuffer.wrap(buffer));
		}
		
		public virtual void  addConnectionListener(ITransportConnectionListener listener)
		{
			//synchronized(listeners) {
			//listenersLock.writeLock().lock_Renamed();
            lock(listeners)
            {
			    listeners.Add(listener);
            }
			//listenersLock.writeLock().unlock();
		}
		
		public virtual void  delConnectionListener(ITransportConnectionListener listener)
		{
            lock(listeners)
            {
			    listeners.Remove(listener);
            }
		}
		
		public virtual void  fireConnectedEvent()
		{
			lock(listeners) 
            {
				foreach(ITransportConnectionListener listener in listeners)
				{
					listener.onConnected(this);
				}
			}
		}
		
		public virtual void  fireDisconnectedEvent()
		{
			lock(listeners) 
            {
				foreach(ITransportConnectionListener listener in listeners)
				{
					listener.onDisconnected(this);
				}
			}
		}
		
		protected internal virtual bool processReceivedCallMessage(MessageEnvelope message)
		{
			bool result = false;
		    if (message != null)
		    {
			    if (currentCallMessageId.Equals(message.Id))
			    {
				    currentCallMessage = message;					    
				    result = true;
			    }
		    }
            if(result) 
            {
                callLockEvent.Set();
            }
			
			if (!result)
			{
				AsyncCallManager mgr = socketFactory.getTransportFactory().AsyncCallManager;
				AsyncCallItem callAsyncResult = mgr.getAsyncCall(message);
				if (callAsyncResult != null)
				{
					result = true;
					if (callAsyncResult.Listener != null)
					{
						callAsyncResult.Listener.onCallResult(callAsyncResult.Request, message);
					}
				}
			}
			
			return result;
		}
		
		protected internal virtual void  doProcessReceivedData(MessageEnvelope message, Transport forTransport)
		{
			if (message.Body.isAliveRequestSelected())
				return ;
			bool doProcessListeners = !forTransport.processReceivedCallMessage(message);
			
			if (doProcessListeners)
			{
                lock(readers)
                {
				    bool handled = false;
				    foreach(ITransportReader reader in readers)
				    {
					    handled = reader.onReceive(message, forTransport);
				    }
				    if (!handled && this.unhanledReader != null)
					    this.unhanledReader.onReceive(message, forTransport);
                }
			}
		}
		
		protected internal virtual void  doProcessReceivedData(ByteBuffer packet, Transport forTransport)
		{
			MessageEnvelope message = messageCoder.decode(packet);
			doProcessReceivedData(message, forTransport);
		}
		
		protected internal virtual void  fireReceivedData(ByteBuffer packet)
		{
			doProcessReceivedData(packet, this);
		}
		
		public virtual MessageEnvelope call(MessageEnvelope message, int timeout)
		{
			MessageEnvelope result = null;
			lock (this)
			{
				currentCallMessage = null;
				currentCallMessageId = message.Id;
			    sendAsync(message);
		        callLockEvent.WaitOne(timeout*1000,false);
				result = currentCallMessage;
				currentCallMessageId = " -none- ";
			}
			if (result == null)
			{
				throw new TimeoutException("Call by RPC-style message timeout in " + this + "!");
			}
		    return result;
		}
		
		public virtual MessageEnvelope call(MessageEnvelope message)
		{
			return this.call(message, 120); // By default timeout is 2 min
		}
		
		public virtual void  callAsync(MessageEnvelope message, ITransportCallListener listener, int timeout)
		{
			AsyncCallManager mgr = socketFactory.getTransportFactory().AsyncCallManager;
			try
			{
				mgr.storeRequest(message, listener, timeout);
				sendAsync(message);
			}
			catch (System.Exception ex)
			{
				mgr.getAsyncCall(message.Id);
				throw ex;
			}
		}
		
		public virtual void  callAsync(MessageEnvelope message, ITransportCallListener listener)
		{
			this.callAsync(message, listener, 120); // By default timeout is 2 min
		}
		
		public virtual Uri getAddr()
		{
			return addr;
		}
		
		protected internal virtual void  setAddr(Uri addr)
		{
			this.addr = addr;
		}
		
		public virtual bool isAvailable()
		{
            lock(addr)
            {
			    //bool result = Socket != null && Socket.isOpen() && Socket.isConnected();
			    //return result;
                return false;
            }
		}
		
		~Transport()
		{
			close();
		}
		
		public virtual void  addReader(ITransportReader reader)
		{
			//readersLock.writeLock().lock_Renamed();
            lock(readers)
            {
			    readers.Add(reader);
            }
		}
		
		public virtual void  delReader(ITransportReader reader)
		{
            lock(readers)
            {
			    readers.Remove(reader);
			}
		}
	}
}